src/bio/bio_monitor.c

```diff
@@ -281,27 +281,68 @@ get_spdk_identify_ctrlr_completion(struct spdk_bdev_io *bdev_io, bool success,
 }
 
 static void
-populate_health_stats(struct spdk_nvme_health_information_page *page,
-		      struct nvme_stats *dev_state)
+populate_health_stats(struct bio_dev_health *bdh)
 {
-	union spdk_nvme_critical_warning_state	cw = page->critical_warning;
-
-	dev_state->warn_temp_time = page->warning_temp_time;
-	dev_state->crit_temp_time = page->critical_temp_time;
-	dev_state->ctrl_busy_time = page->controller_busy_time[0];
-	dev_state->power_cycles = page->power_cycles[0];
-	dev_state->power_on_hours = page->power_on_hours[0];
-	dev_state->unsafe_shutdowns = page->unsafe_shutdowns[0];
-	dev_state->media_errs = page->media_errors[0];
+	struct spdk_nvme_health_information_page	*page;
+	struct nvme_stats				*dev_state;
+	union spdk_nvme_critical_warning_state		cw;
+
+	page		= bdh->bdh_health_buf;
+	cw		= page->critical_warning;
+	dev_state	= &bdh->bdh_health_state;
+
+	/** commands */
+	d_tm_set_counter(bdh->bdh_du_written, page->data_units_written[0]);
+	d_tm_set_counter(bdh->bdh_du_read, page->data_units_read[0]);
+	d_tm_set_counter(bdh->bdh_write_cmds, page->host_write_commands[0]);
+	d_tm_set_counter(bdh->bdh_read_cmds, page->host_read_commands[0]);
+	dev_state->ctrl_busy_time	= page->controller_busy_time[0];
+	d_tm_set_counter(bdh->bdh_ctrl_busy_time,
+			 page->controller_busy_time[0]);
+	dev_state->media_errs		= page->media_errors[0];
+	d_tm_set_counter(bdh->bdh_media_errs, page->media_errors[0]);
+
+	dev_state->power_cycles		= page->power_cycles[0];
+	d_tm_set_counter(bdh->bdh_power_cycles, page->power_cycles[0]);
+	dev_state->power_on_hours	= page->power_on_hours[0];
+	d_tm_set_counter(bdh->bdh_power_on_hours, page->power_on_hours[0]);
+	dev_state->unsafe_shutdowns	= page->unsafe_shutdowns[0];
+	d_tm_set_counter(bdh->bdh_unsafe_shutdowns,
+			 page->unsafe_shutdowns[0]);
+
+	/** temperature */
+	dev_state->warn_temp_time	= page->warning_temp_time;
+	d_tm_set_counter(bdh->bdh_temp_warn_time, page->warning_temp_time);
+	dev_state->crit_temp_time	= page->critical_temp_time;
+	d_tm_set_counter(bdh->bdh_temp_crit_time, page->critical_temp_time);
+	dev_state->temperature		= page->temperature;
+	d_tm_set_counter(bdh->bdh_temp, page->temperature);
+	dev_state->temp_warn		= cw.bits.temperature ? true : false;
+	d_tm_set_counter(bdh->bdh_temp_warn, dev_state->temp_warn);
+
+	/** reliability */
+	d_tm_set_counter(bdh->bdh_avail_spare, page->available_spare);
+	d_tm_set_counter(bdh->bdh_avail_spare_thres,
+			 page->available_spare_threshold);
+	dev_state->avail_spare_warn	= cw.bits.available_spare ? true
+								  : false;
+	d_tm_set_counter(bdh->bdh_avail_spare_warn,
+			 dev_state->avail_spare_warn);
+	dev_state->dev_reliability_warn	= cw.bits.device_reliability ? true
+								     : false;
+	d_tm_set_counter(bdh->bdh_reliability_warn,
+			 dev_state->dev_reliability_warn);
```

Comment:
Shouldn't most of these be gauges? Counters are supposed to track a value that only increases. Most if not all of these could fluctuate up or down or be 0/1, right?

Reply:
I did not use gauge because it is pointless to track min/max/avg/stddev for those. Maybe we should have different types, for instance:
- D_TM_GAUGE with just the latest value, useful for the use case here
- D_TM_GAUGE_MMA with min/max/average
- D_TM_GAUGE_STD with min/max/average and standard deviation
unless we prefer to add flags at creation time?

Reply:
Yeah, maybe just D_TM_GAUGE (current behavior) and D_TM_GAUGE_SIMPLE (latest value only)? I guess the question is which should be the default... Maybe D_TM_GAUGE (simple) and D_TM_GAUGE_STATS (all the bells & whistles)? I don't know which use case is more common.

Reply:
My instinct would be to use D_TM_GAUGE as the simple type, and D_TM_GAUGE_STATS as the advanced type, myself. It would help to know which case is more commonly wanted though.